tests/cpp/test_indexing.cpp

```diff
@@ -4850,6 +4851,233 @@ TEST_F(ContigIndexingTest, ConcretizedBroadcastMerge) {
+
+  inlineMost();
+
+  tv2->axis(0)->parallelize(ParallelType::BIDx);
+  tv2->axis(1)->parallelize(ParallelType::Unswitch);
+  tv2->axis(3)->parallelize(ParallelType::TIDx);
+
+  struct GetReference : AbstractGetReference {
+    GetReference(const TensorIndexer& indexer, const IdModel& id_model)
+        : AbstractGetReference(indexer, id_model) {}
+
+    // The unswitch predicate should look like:
+    //
+    // (((blockIdx.x * 4 + 0) * 8 + 0) * 128 + threadId.x >= 0 &&
+    // (((blockIdx.x * 4 + 3) * 8 + 7) * 128 + threadId.x <
+    // tv0.logical_size[0] * tv0.logical_size[1]
+    Val* getOuterPredicate(TensorView* tv) const override {
+      std::vector<Val*> loop_indices = getLoopIndices(tv, indexer_);
+      auto start_idx = addExpr(
+          mulExpr(
+              IrBuilder::addExpr(
+                  mulExpr(
+                      IrBuilder::addExpr(
+                          mulExpr(loop_indices.at(0), tv->axis(1)->extent()),
+                          tv->fusion()->zeroVal()),
+                      tv->axis(2)->extent()),
+                  tv->fusion()->zeroVal()),
+              tv->axis(3)->extent()),
+          loop_indices.at(3));
+      auto stop_idx = addExpr(
+          mulExpr(
+              IrBuilder::addExpr(
+                  mulExpr(
+                      IrBuilder::addExpr(
+                          mulExpr(loop_indices.at(0), tv->axis(1)->extent()),
+                          subExpr(
+                              tv->axis(1)->extent(), tv->fusion()->oneVal())),
+                      tv->axis(2)->extent()),
+                  subExpr(tv->axis(2)->extent(), tv->fusion()->oneVal())),
+              tv->axis(3)->extent()),
+          loop_indices.at(3));
+
+      return andExpr(
+          geExpr(start_idx, tv->fusion()->zeroVal()),
+          ltExpr(
+              stop_idx,
+              mulExpr(
+                  tv->getLogicalDomain().at(0)->extent(),
+                  tv->getLogicalDomain().at(1)->extent())));
+    }
+  };
+
+  PredicateIndexValidator<GetReference>::validate(&fusion, true);
+}
+
+// Make sure non-divisible split to prevent contig predicate indexing
+TEST_F(ContigPredicateIndexingTest, NonDivisibleSplit1) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeSymbolicTensor(2);
+  fusion.addInput(tv0);
+
+  // [I0, I1] -> [I0/5, 5, I1]
+  auto tv1 = reshape(tv0, {10, 20}, {2, 5, 20});
+  fusion.addOutput(tv1);
+
+  auto tv2 = set(tv0);
+  fusion.addOutput(tv2);
+
+  // Merge first one of the split outputs with the innermost. Since
+  // the split for the reshape is guaranteed to be divisible, the
+  // output domain of the last merge should be the one to predicate
+  tv1->merge(1, 2);
+  tv1->merge(0, 1);
+
+  // While the transformations are the same as tv1, nvFuser doesn't
+  // know the corresponding split is actually divisible, so it's
+  // considered non-divisible. The resulting predicaion should be done
+  // with each of the two logical domains.
```

Comment:
Off-topic but it's too bad we can't attest that the splits done during a reshape are always divisible. Actually I guess the existence of a split along a root->logical path in a `ViewOp` consumer could actually be used to determine that the `ExprGroup` of the split is divisible...

Reply:
We do use that information by collecting all splits used in reshape transformations. Here, it is a rather synthetic case. The reshape split is:

```
Outer split: iS17{( (( (( getMetaData(T0) )).logical_size ))[0] )}rf by factor 2 -> iS5{2}rf, iS6{( ceilDiv(( (( (( getMetaData(T0) )).logical_size ))[0] ), 2) )}rf
```

The split of tv2 is:

```
Split: iS7{i0} by factor 5 -> iS11{( ceilDiv(i0, 5) )}, iS12{5}
```

As you can see, they are effectively the same operation, but our analysis is not comprehensive enough to detect the equality. I'm not sure if it's worthwhile to extend the divisibility analysis unless there's any actual use case.

Reply:
Ah that's right. The `getAllDivisibleSplits` lowering pass does what I was saying and fills the `divisible_splits_` set that's used here. However as you say proving a divisible outer split is equal to the corresponding inner split is another matter.

Reply:
Added a clarifying comment.